### PR TITLE
Introduce wp_trigger_error()

### PR DESCRIPTION
**Pull request by azaozz**
Trac ticket: https://core.trac.wordpress.org/ticket/57686

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

**Comment by azaozz** [comment]
TODO/TBD: Wondering if it's worth it to add a simple backtrace when `WP_DEVELOPMENT_MODE` is true. Tested with only printing function names, no args as they can be huge sometimes. The code for it would be like:
```
	if ( WP_DEVELOPMENT_MODE ) {
		// Output a simple backtrace.
		ob_start();
		debug_print_backtrace( DEBUG_BACKTRACE_IGNORE_ARGS );
		$backtrace = ob_get_contents();
		ob_end_clean();

		// This may be called very early. See above about use of `__()`.
		if ( function_exists( 'esc_html' ) ) {
			$backtrace = esc_html( $backtrace );
		}

		// Space it a bit for better readability.
		$backtrace = str_replace( "\n", "\n\n", $backtrace );

		echo '<p style="white-space: pre-wrap;">';
		echo $backtrace;
		echo '</p>';
	}
```

File: src/wp-includes/functions.php [hunk 1]
```diff
@@ -6000,42 +6000,96 @@ function _doing_it_wrong( $function_name, $message, $version ) {
...
+	 */
+	if ( true !== apply_filters( 'wp_trigger_error', true, $where, $message, $error_level ) ) {
+		return;
+	}
+
+	if ( ! WP_DEBUG ) {
+		return;
+	}
+
+	if ( function_exists( '__' ) ) {
```
Comment (hellofromtonya) [inline src/wp-includes/functions.php:6054]: Why translate PHP errors / warnings / notices / deprecations?

Comment (azaozz) [reply src/wp-includes/functions.php:6054]: Good question. These are translated in doing_it_wrong, was just using that code as example. 
As these errors are intended only for developers and not for the end users, translation is probably not needed.

**Comment by hellofromtonya** [comment]
Closing in favor of a wrapper approach in https://github.com/WordPress/wordpress-develop/pull/5144, that was committed instead.

**Pull request closed without merging**